staff/views/projects.py

```diff
@@ -46,16 +47,12 @@ def dispatch(self, request, *args, **kwargs):
         return super().dispatch(request, *args, **kwargs)
 
     def form_valid(self, form):
-        roles = form.cleaned_data["roles"]
-        users = form.cleaned_data["users"]
-
-        with transaction.atomic():
-            for user in users:
-                self.project.memberships.create(
-                    user=user,
-                    created_by=self.request.user,
-                    roles=roles,
-                )
+        members.add(
```

Comment:
Related to this, adding an interactive user to a project doesn't call this command, so no audit log entry is generated. I don't know if there are any other places where we change project membership that I haven't tested?

Reply:
> How will future developers know to use this command

They won't until we move to using commands/a domain layer across the codebase, but we've been using them in interactive and releases for 2y+ wtihout incident.  When I've come across code using them it's clear to me to use those.  It's by no means perfect but we don't have much choice until we move everything over.

Reply:
In addition to their other spooky-action-at-a-distance-confusion issues, IMO signals are by design not reliable enough to implement an audit trail that is a critical part of our business logic.

Audit events *must* be committed in the same transaction as the state change they audit, otherwise we cannot trust them.  AIUI, signals cannot do this cleanly. Django famously rejected transaction safe signals a [long time ago](https://code.djangoproject.com/ticket/14051).  There are some workarounds, like using [`transaction.on_commit`](https://docs.djangoproject.com/en/4.2/topics/db/transactions/#performing-actions-after-commit) to queue actions once the current transaction is committed successfully. But it terms of *writes to the db*, that is not at all the same thing, as the second write could fail.

I think signals are designed to solve different kind of problems to what we are trying to do here.

I think we need to control this explicitly, and a good way to do that IME is to wrap the state change in an application domain specific api call/layer, like George has done here. This approach has other benefits too, but I digress.

However, I do think Lucy's concern about folks not knowing and bypassing by using `Users.models.create()` is somewhat valid, given the importance of the audit trail.

One option to fix this issue could be to disable the default `project.membership.create()` function, have it raise an exception. IIRC, this can be done via custom manager[1] for the Membership model that raises on those methods. I'm hazy on the details, and there may be a better way these days, but I think something like that should be doable? Probably as a follow on PR?


[1] @ghickman: I am a heretic! But hear me out: managers that just raise are the *one* valid use of them I have found! :rofl:  And the alternative is db triggers, so... :D

Reply:
Done in 9930a5e3. The highlights are that I've disabled ProjectMembership's default `create()` but also added a, currently unused, `override` kwarg in case we need to go around the command function.  I've also removed the ProjectMembershipFactory in favour of a fixture now that much of the utility of the factory would need extra code.